src/upgrade.py

```diff
@@ -0,0 +1,151 @@
+        self.pre_upgrade_check()
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        default_message = "Pre-upgrade check failed and cannot safely upgrade"
+        try:
+            if not self.client.members_broadcasting or not len(self.client.server_members) == len(
+                self.charm.cluster.peer_units
+            ):
+                raise ClusterNotReadyError(
+                    message=default_message,
+                    cause="Not all application units are connected and broadcasting in the quorum",
+                )
+
+            if self.client.members_syncing:
+                raise ClusterNotReadyError(
+                    message=default_message, cause="Some quorum members are syncing data"
+                )
+
+            if not self.charm.cluster.stable:
+                raise ClusterNotReadyError(
+                    message=default_message, cause="Charm has not finished initialising"
+                )
+
+        except QuorumLeaderNotFoundError:
+            raise ClusterNotReadyError(message=default_message, cause="Quorum leader not found")
+        except ConnectionClosedError:
+            raise ClusterNotReadyError(
+                message=default_message, cause="Unable to connect to the cluster"
+            )
+        except Exception as e:
+            logger.debug(str(e))
+            raise ClusterNotReadyError(message=default_message, cause="Unknown error")
+
+    @override
+    def build_upgrade_stack(self) -> list[int]:
+        upgrade_stack = []
+        for unit in self.charm.cluster.peer_units:
+            config = self.charm.cluster.unit_config(unit=unit)
+
+            # upgrade quorum leader last
+            if config["host"] == self.client.leader:
+                upgrade_stack.insert(0, int(config["unit_id"]))
+            else:
+                upgrade_stack.append(int(config["unit_id"]))
+
+        return upgrade_stack
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        logger.critical(
+            "\n".join(
+                [
+                    "Unit failed to upgrade and requires manual rollback to previous stable version.",
+                    "    1. Re-run `pre-upgrade-check` action on the leader unit to enter 'recovery' state",
+                    "    2. Run `juju refresh` to the previously deployed charm revision",
+                ]
+            )
+        )
+        return
+
+    @override
+    def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
+        self.charm.snap.stop_snap_service()
+
+        if not self.charm.snap.install():
+            logger.error("Unable to install ZooKeeper Snap")
+            self.set_unit_failed()
+            return
+
+        logger.info(f"{self.charm.unit.name} upgrading service...")
+        self.charm.snap.restart_snap_service()
+
+        try:
+            logger.debug("Running post-upgrade check...")
+            self.post_upgrade_check()
```

Comment:
question: Wouldn't the normal execution of the charm already catch errors on the cluster? We don't have a way of knowing why the cluster failed, so the idea here is to assume that if it failed, it was because of the upgrade. Is this thought correct?